src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -677,6 +742,234 @@ impl Crumbable for crate::Block<Ast> {
+                    (Except(pat) , PatternMatchCrumb::Except  ) => {
+                        pat.elem = pat.elem.clone();
+                        pattern  = Rc::make_mut(&mut pat.elem);
+                    },
+                    (Tag(pat), PatternMatchCrumb::Tag) => {
+                        pat.elem = pat.elem.clone();
+                        pattern  = Rc::make_mut(&mut pat.elem);
+                    },
+                    (Cls(pat), PatternMatchCrumb::Cls) => {
+                        pat.elem = pat.elem.clone();
+                        pattern  = Rc::make_mut(&mut pat.elem);
+                    },
+                    (Or(pat), PatternMatchCrumb::Or) => match &mut pat.elem {
+                        crate::Switch::Left (pat) => {
+                            pat.value = pat.value.clone();
+                            pattern   = Rc::make_mut(&mut pat.value);
+                        },
+                        crate::Switch::Right(pat) => {
+                            pat.value = pat.value.clone();
+                            pattern   = Rc::make_mut(&mut pat.value);
+                        },
+                    },
+                    (Seq(pat), PatternMatchCrumb::Seq{right}) => {
+                        if *right {
+                            pat.elem.1 = pat.elem.1.clone();
+                            pattern    = Rc::make_mut(&mut pat.elem.1);
+                        } else {
+                            pat.elem.0 = pat.elem.0.clone();
+                            pattern    = Rc::make_mut(&mut pat.elem.0);
+                        }
+                    },
+                    (Many(pat), PatternMatchCrumb::Many{index}) => {
+                        let elem = pat.elem.get_mut_or_err(*index,"elem")?;
+                        *elem    = elem.clone();
+                        pattern  = Rc::make_mut(elem);
+                    },
+                    _ => return missing("crumb"),
+                }
+            }
+            Ok(result)
+        }
+
+        let mut new_self = self.clone();
+
+        match crumb {
+            MatchCrumb::Pfx {val} => match new_self.pfx {
+                None      => return Err(NotPresent("prefix".into()).into()),
+                Some(pat) => new_self.pfx = Some(set_pattern(&pat,val,new_ast)?),
+            }
+
+            MatchCrumb::Segs{index,val:SegmentMatchCrumb::Body{val}} => {
+                let mut seg = new_self.segs.get_mut_or_err(*index,"segment")?;
+                seg.body    = set_pattern(&seg.body,val,new_ast)?
+            },
+            MatchCrumb::Segs{index,val:SegmentMatchCrumb::Head} => {
+                let mut seg = new_self.segs.get_mut_or_err(*index,"segment")?;
+                seg.head    = new_ast
+            }
+        }
+        Ok(new_self)
+
+    }
+
+    fn iter_subcrumbs<'a>(&'a self) -> Box<dyn Iterator<Item=Self::Crumb> + 'a> {
+
+        fn iter_pat(pat:&MacroPatternMatch<Shifted<Ast>>) -> Vec<Vec<PatternMatchCrumb>> {
```

Comment:
Could we have this return `Iterator<Item=Vec<PatternMatchCrumb>>` (in box if necessary?)

All `iter_subcrumbs` provide iterator-based API and that leads user to assumption that he will pay cost of traversing continually through traversal (and that cost for non-visisted subtrees is not paid).

Not a blocker (can be fixed later if ever relevant) but if feasible, would be nice.